vlasovsolver/cuda_acc_map_kernel.cu

```diff
@@ -50,144 +71,313 @@ __global__ void acceleration_1
   int bdsw3
 )
 {
-  //int index = threadIdx.x + blockIdx.x*blockDim.x;
-  //if(index == 0)
-  {
-    //if(totalColumns > 256)
-    //  printf("totalColumns = %d;\n", totalColumns);
-    int column = threadIdx.x + blockIdx.x * blockDim.x;
-    if(column < totalColumns)
-    //for( uint column=0; column < totalColumns; column++)
-    {
+   //int index = threadIdx.x + blockIdx.x*blockDim.x;
+   //if(totalColumns > 256)
+   //  printf("totalColumns = %d;\n", totalColumns);
+   //int column = threadIdx.x + blockIdx.x * blockDim.x;
+   //if(column < totalColumns)
+   //for( uint column=0; column < totalColumns; column++)
+
+#ifdef CUDA_REALF
+
+   // Optimization idea for future: reverse-sort columns based on
+   // block counts
```

Comment:
Do you mean this optimization to minimize the computational divergence within one block? Because columns of similar length will be grouped together?

I suppose the easiest place to do this is in column construction, and an approximate bucket-sort would be sufficient for the purpose? Because that could be done quickly on the GPU in the same vein.

Reply:
So currently, columns are divvied amongs cudablocks with a simple modulo-style approach. Now if one cudablock happens to get several long columns and others get short columns, there can be imbalance between cudablock work sizes.

If we reverse-sort columns based on block counts, the low index cudablocks would get N columns starting from the shortest ones, and the high index cudablocks would get N-1 columns from the larger end, thus balancing the load a bit. And yes an approximate sort would be fine.

Another approach I've been thinking about would be to split the actual blocks evenly across cudablocks as it's only the intersection/_min/_max calculations (L111-L129) which would be repeated. With smart indexing we'd still access continuous memory extents.

Reply:
Instead of divvying columns manually, I now allow the driver to pass them to blocks as resources become available. At least in WID8 SPF it gave a ~30% speed increase!